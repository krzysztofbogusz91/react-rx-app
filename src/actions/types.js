export const ADD_TO_CHOSEN = 'ADD_TO_CHOSEN';
export const FETCH_HEROES_SUCCESS = 'FETCH_HEROES_SUCCESS';
export const REMOVE_FROM_CHOSEN = 'REMOVE_FROM_CHOSEN';
