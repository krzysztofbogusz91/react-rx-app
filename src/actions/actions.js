import 'babel-polyfill';

export const ADD_TO_CHOSEN = 'ADD_TO_CHOSEN';
export const FETCH_HEROES_SUCCESS = 'FETCH_HEROES_SUCCESS';
export const REMOVE_FROM_CHOSEN = 'REMOVE_FROM_CHOSEN';

export const addHero = (hero) => ({ type: ADD_TO_CHOSEN, hero });
export const removeHero = (hero) => ({ type: REMOVE_FROM_CHOSEN, hero });
export const getHeros = (heros) => ({ type: FETCH_HEROES_SUCCESS, heros });

const getWorlds = (list) => Promise.all(list.map(async (hero) => {
  const world = await fetch(hero.homeworld).then((resp) => resp.json());
  return { ...hero, homeworld: world.name };
}));

const getSpecies = (list) => Promise.all(list.map(async (hero) => {
  const species = await (Promise.all(hero.species.map(async (spicie) => {
    const spi = await fetch(spicie).then((resp) => resp.json());
    return spi.name;
  })));

  return { ...hero, species: [...species] };
}));
export const getHeroesAction = () => (dispatch) => fetch('https://swapi.co/api/people/')
  .then((resp) => resp.json())
  .then(async (listOfHeroes) => {
    const herosWithHome = await getWorlds(listOfHeroes.results);
    const herosWithSpecies = await getSpecies(herosWithHome);
    dispatch(getHeros(herosWithSpecies
      .map((hero, id) => ({ ...hero, img: `https://i.pravatar.cc/200?img=${id + 1}` }))));
  }, (err) => err)
  .catch((err) => err);
