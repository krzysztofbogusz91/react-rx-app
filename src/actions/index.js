import 'babel-polyfill';
import { ADD_TO_CHOSEN, REMOVE_FROM_CHOSEN, FETCH_HEROES_SUCCESS } from './types';

export const addHero = (hero) => ({ type: ADD_TO_CHOSEN, hero });
export const removeHero = (hero) => ({ type: REMOVE_FROM_CHOSEN, hero });
export const getHeros = (heros) => ({ type: FETCH_HEROES_SUCCESS, heros });


const getWorlds = (list) => Promise.all(list.map(async (hero) => {
  const world = await fetch(hero.homeworld).then((resp) => resp.json());
  return { ...hero, homeworld: world.name };
}));

const getSpecies = (list) => Promise.all(list.map(async (hero) => {
  const species = await (Promise.all(hero.species.map(async (spicie) => {
    const spi = await fetch(spicie).then((resp) => resp.json());
    return spi.name;
  })));

  return { ...hero, species: [...species] };
}));

export const getHeroesAction = () => (dispatch) => fetch('https://swapi.co/api/people/')
  .then((resp) => resp.json())
  .then(async (listOfHeroes) => {
    const herosWithHome = await getWorlds(listOfHeroes.results);
    const herosWithSpecies = await getSpecies(herosWithHome);
    dispatch(getHeros(herosWithSpecies
      .map((hero, id) => ({ ...hero, img: `https://i.pravatar.cc/200?img=${id + 1}` }))));
  }, (err) => err)
  .catch((err) => err);
