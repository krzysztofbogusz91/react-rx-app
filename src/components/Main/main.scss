@import '~shared/styles/index.scss';

.main_view {
    width: 90%;
}
.main_view_list_container {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: center;
}
