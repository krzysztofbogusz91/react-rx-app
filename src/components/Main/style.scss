@import '~shared/styles/index.scss';

.header{
  margin: 0;
  font-family:  Verdana, sans-serif;
}
