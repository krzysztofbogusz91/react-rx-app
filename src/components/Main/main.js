import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ChosenList from '../ChosenList/chosen-list';
import HeroList from '../HeroList/hero-list';
import SearchBar from '../SearchBar/search-bar';
import Header from '../Header/header';
import './main.scss';
import { getHeroesAction } from '../../actions';
import { Subject } from 'rxjs';
import { debounceTime } from 'rxjs/operators';

class Main extends Component {
  constructor() {
    super();
    this.search$ = new Subject();
    this.search = this.search$.asObservable().pipe(
      debounceTime(500),
    );
  }

  componentDidMount() {
    const { getHeroes } = this.props;
    getHeroes().subscribe();
    this.search.subscribe((text) => {
      console.log(text);
    });
  }

  onSearch = (e) => {
    this.search$.next(e.target.value);
  }

  render() {
    const { listOfHeroes, team } = this.props;
    return (
      <div className="main_view">
        <Header />
        <SearchBar onSearch={this.onSearch} />
        <div className="main_view_list_container">
          <HeroList heroesList={listOfHeroes} />
          <ChosenList chosenList={team} />
        </div>
      </div>
    );
  }
}

Main.propTypes = {
  listOfHeroes: PropTypes.arrayOf(PropTypes.object),
  team: PropTypes.arrayOf(PropTypes.object),
  getHeroes: PropTypes.func,
};
Main.defaultProps = {
  listOfHeroes: [],
  team: [],
  getHeroes: null,
};

const mapStateToProps = (state) => ({ team: state.team, listOfHeroes: state.list });

const mapDispatchToProps = (dispatch) => ({
  getHeroes: () => dispatch(getHeroesAction()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Main);
