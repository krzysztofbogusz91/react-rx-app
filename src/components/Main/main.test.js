import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import TestRenderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import Main from './main';
import { getHeroesAction } from '../../actions';

const { act, create } = TestRenderer;
const mockStore = configureStore([thunk]);

describe('Main', () => {
  let container = null;
  let component = null;
  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });
  it('should get data from api', () => {
    const fakeResponse = [{ name: 'Person 1', species: [] }, { name: 'Person 2', species: [] }];

    act(() => {
      const store = mockStore({ team: [], list: [...fakeResponse] });
      component = create(
        <Provider store={store}>
          <Main team={[]} listOfHeroes={[]} />
        </Provider>,
      );
      getHeroesAction();
    });
    expect(component.toJSON()).toMatchSnapshot();
  });
});
