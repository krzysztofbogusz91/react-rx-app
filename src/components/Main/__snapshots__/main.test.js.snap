// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should get data from api 1`] = `
<div
  className="main_view"
>
  <div>
    <h1
      className="header"
    >
       Gather the team before setting off: 
    </h1>
  </div>
  <div
    className="search_bar"
  >
    <div>
      <label
        htmlFor="hero-inputs"
      >
        Search for heroes: 
        <input />
      </label>
    </div>
  </div>
  <div
    className="main_view_list_container"
  >
    <div
      className="hero_list_container"
    >
      <ul
        className="hero_list"
      >
        <li
          className="hero"
        >
          <div
            className="img_container"
          >
            <div
              className="hero_img"
            >
              <img
                alt="hero_avatar"
              />
            </div>
          </div>
          <div
            className="info_container"
          >
            <div
              className="hero_name"
            >
              Name: 
              Person 1
            </div>
            <div>
              <span>
                Home World: 
                 
              </span>
            </div>
            <div>
              Hair Color: 
            </div>
            <div>
              Specie:
              <ul
                className="spi_list"
              />
            </div>
            <div>
              <button
                className="btn_add"
                href="#"
                onClick={[Function]}
                type="button"
              >
                 Add to list...
              </button>
            </div>
          </div>
        </li>
        <li
          className="hero"
        >
          <div
            className="img_container"
          >
            <div
              className="hero_img"
            >
              <img
                alt="hero_avatar"
              />
            </div>
          </div>
          <div
            className="info_container"
          >
            <div
              className="hero_name"
            >
              Name: 
              Person 2
            </div>
            <div>
              <span>
                Home World: 
                 
              </span>
            </div>
            <div>
              Hair Color: 
            </div>
            <div>
              Specie:
              <ul
                className="spi_list"
              />
            </div>
            <div>
              <button
                className="btn_add"
                href="#"
                onClick={[Function]}
                type="button"
              >
                 Add to list...
              </button>
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div
      className="chosen_list_container"
    >
      <ul
        className="chosen_list"
      >
        <li>
          <h2
            style={
              Object {
                "marginTop": "20px",
              }
            }
          >
            No one chosen...
          </h2>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
